pkg/plugins/manager/signature/manifestverifier/verifier.go

```diff
@@ -111,20 +152,73 @@ func (pmv *ManifestVerifier) GetPublicKey(keyID string) (string, error) {
 	}()
 
 	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
-		return "", err
+		return err
 	}
 
 	if len(data.Items) == 0 {
-		return "", errors.New("missing public key")
+		return errors.New("missing public key")
+	}
+
+	cachedKeys, err := pmv.kv.ListKeys(ctx)
+	if err != nil {
+		return err
 	}
 
+	shouldKeep := make(map[string]bool)
 	for _, key := range data.Items {
-		pmv.publicKeys[key.KeyID] = key
+		err = pmv.kv.Set(ctx, key.KeyID, key.PublicKey)
+		if err != nil {
+			return err
+		}
+		shouldKeep[key.KeyID] = true
+	}
+
+	// Delete keys that are no longer in the API
+	for _, key := range cachedKeys {
+		if !shouldKeep[key] {
+			err = pmv.kv.Del(ctx, key)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Update the last updated timestamp
+	pmv.kv.GetLastUpdated(ctx)
+
+	return nil
+}
+
+// getPublicKey loads public keys from:
+//   - The hard-coded value if the feature flag is not enabled.
+//   - A cached value from kv storage if it has been already retrieved. This cache is populated from the grafana.com API.
+func (pmv *ManifestVerifier) GetPublicKey(keyID string) (string, error) {
+	if !pmv.featureEnabled() {
+		return publicKeyText, nil
 	}
 
-	key, exist = pmv.publicKeys[keyID]
+	pmv.lock.Lock()
+	defer pmv.lock.Unlock()
+
+	ctx := context.Background()
```

Comment:
This now requires a `context.Context` due to the kvstore, I think it's better to now change the signature to accept a `context.Context` as first argument rather than creating a `context.Background()` here.

I have noticed the call chain for this one is a bit long until we get to a usable `context.Context`

(`GetPublicKey` -> `Signature.validateManifest` -> `Signature.readPluginManifest` -> `Signature.Calculate`)

but I would still do it, as it's the only usage apart from tests.

Alternatively, I would at least add some sort of timeout with `context.WithTimeout`, but in my opinion it's worth passing the whole context around.